### PR TITLE
Improve global error detection

### PR DESCRIPTION
**Pull request by leonklingele**
Only treat global error variables as such if they have an "err" or "Err" prefix
and implement the error interface.

Close https://github.com/leighmcculloch/gochecknoglobals/issues/5

### CODE REVIEW

**Comment by hurrycaner** [comment]
I would like to suggest making that prefix configurable, i would like to use the error interface implementation feature, but not that err or Err prefixes

**Comment by leighmcculloch** [comment]
This looks good to me, but I'm a little concerned with the fact that this is a breaking change potentially as it reduces the surface area of what is allowed. It's okay for us to increase the surface area of what is allowed but reducing it may break existing builds. 

Maybe that's okay. Does anyone else have thoughts on this?

**Comment by leighmcculloch** [comment]
@hurrycaner the idea to use the err prefix for non error globals sounds out of scope of this tool.

**Comment by leonklingele** [comment]
Hey @leighmcculloch, can this be merged? 😊

**Review by leighmcculloch — COMMENTED** [review]
This looks good to me. There's one test case I'd like to add, to make sure we don't have a gap there. See below.

File: checknoglobals/testdata/src/7/code.go [hunk 1]
```diff
@@ -5,30 +5,55 @@ import (
...
 var declaredErr error = errors.New("declaredErr") // want "declaredErr is a global variable"
 var errDeclared error = errors.New("errDeclared")
+var declaredErrFunc error = func() error { return errors.New("declaredErrFunc") }() // want "declaredErrFunc is a global variable"
+var errDeclaredFunc error = func() error { return errors.New("errDeclaredFunc") }()
 
 type customError struct{ e string }
 
 func (e *customError) Error() string { return e.e }
+
+func (e *customError) SomeOtherMethod() string { return e.e }
```
[the remark this answers is not in the document]
Comment (leighmcculloch) [reply checknoglobals/testdata/src/7/code.go:59]: Could we also add an error type that isn't a struct and implements `error`? Maybe a `type customStringError string`.

[the remark this answers is not in the document]
Comment (leighmcculloch) [reply checknoglobals/testdata/src/7/code.go:63]: Added in 703116572a563755fdcefa33c5d9e6adbf75379b.

File: checknoglobals/check_no_globals.go [hunk 2]
```diff
@@ -86,16 +87,23 @@ func isAllowedSelectorExpression(v *ast.SelectorExpr) bool {
...
 	prefix := "err"
 	if i.IsExported() {
 		prefix = "Err"
 	}
-	return strings.HasPrefix(i.Name, prefix)
+
+	hasErrPrefix := strings.HasPrefix(i.Name, prefix)
+	if !hasErrPrefix {
+		return false
+	}
```
Comment (leighmcculloch) [inline checknoglobals/check_no_globals.go:101]: @leonklingele @bombsimon Do you folks have thoughts on if we should remove the err/Err prefix requirement? Regardless of the name given to the global variable, just checking if it is an error seems reasonable to me.

**Review by leighmcculloch — APPROVED** [review]

**Comment by leighmcculloch** [comment]
Release in v0.2.0.

File: checknoglobals/check_no_globals.go (hunk 2 above)
Comment (leonklingele) [reply checknoglobals/check_no_globals.go]: I'd leave it as-is. We could make it configurable with a default that still enforces the prefix but I don't really see a benefit in doing so.

**Pull request merged**